### DISCUSSION TITLE
Why is Pyright authored in TS?

### DISCUSSION THREAD

**Discussion by mheiber** [microsoft/pyright#2529] [Q&A] [answered]
Why is Pyright authored in TS?

This isn't a criticism, I'm asking so I can learn about the trade-offs.

For other gradual type checkers, I've seen either of these approaches:
- author in the target language so dogfooding can aid language design and/or ease contribution (TypeScript and probably MyPy, Pytype, Typed Racket)
- author in a faster language for performance (Sorbet, Luau, and probably Pyre)

But Pyright seems to have found a third way.

Thanks for sharing your knowledge, and for typed Python.

**Comment by jakebailey** [comment, marked as answer]
@erictraut will probably give a more historical answer, but a current main benefit for end-users is that this allows us to ship pyright/Pylance in places that already support JS:

- VS Code is an Electron app, so we can run python analysis without even having python installed, as you can just reuse the same executable as a node executable by setting an environment variable. This means we can analyze code on remote machines without Python, or on newfangled platforms like macOS arm64 when Python hadn't been compiled there yet, or similar, so long as VS Code is there.
- VS Code in the browser with Pylance on vscode.dev or github.dev is much the same; we're in the browser which is JS, or you have to deal with WASM and that's its own fun. If you saw the vscode.dev/github.dev release annoucements, you'll see that the only languages with full support are TS/JS and Python, because all of that is in TypeScript and effort was spent supporting it. Python itself doesn't do WASM, at least not yet.
- GitHub Actions are natively JS, which means I can piggy back on the same Node install for https://github.com/jakebailey/pyright-action (but, this is mostly moot for the average builder which has node installed anyway).

And, V8 (the actual engine behind node), is really fast, surprisingly so, compared to other interpreted languages.

**Reply by erictraut** [reply]
I think Jake already nicely covered the primary reasons. Here are a few additional thoughts.

Performance was an important consideration. Javascript simply runs much faster than the equivalent Python. (I guess that means you could include Pyright under your second bullet, "author in a faster language for performance".) The CPython developers have made good progress on performance over the past few releases — and you'll likely see more gains in Python 3.11, but certain design considerations intrinsic to Python mean that it will probably never run as fast as Javascript or other JIT'ed languages.

In addition to all of the benefits Jake mentioned, another consideration was my own efficiency as a developer. I find that I'm more productive in TypeScript than I am in other languages including Python. That said, Pyright / Pylance have helped close the gap over the past few years. With strictly-typed Python, I'm now able to write and maintain high-quality Python code much more efficiently than before.
